service/worker/scanner/history/scavenger.go

```diff
@@ -317,6 +332,46 @@ func (s *Scavenger) getPaginationFn(
 	}
 }
 
+func (s *Scavenger) cleanUpWorkflowPassedRetention(
+	ctx context.Context,
+	mutableState *persistencepb.WorkflowMutableState,
+) error {
+	executionInfo := mutableState.GetExecutionInfo()
+	ns, err := s.registry.GetNamespaceByID(namespace.ID(executionInfo.GetNamespaceId()))
+	switch err.(type) {
+	case *serviceerror.NotFound,
+		*serviceerror.NamespaceNotFound:
+		return nil
+	case nil:
+		// continue to delete
+	default:
+		return err
+	}
+
+	retention := ns.Retention()
+	finalUpdateTime := executionInfo.GetLastUpdateTime()
+	ttl := time.Now().UTC().Sub(timestamp.TimeValue(finalUpdateTime))
+	if ttl > 0 && ttl > retention+s.historyDataMinAge() {
+		_, err = s.adminClient.DeleteWorkflowExecution(ctx, &adminservice.DeleteWorkflowExecutionRequest{
+			Namespace: ns.Name().String(),
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: executionInfo.GetWorkflowId(),
+				RunId:      mutableState.GetExecutionState().GetRunId(),
+			},
+		})
+		if err != nil {
+			// This is experimental. Ignoring the error so it will not block the history scavenger.
+			s.logger.Warn("Failed to delete workflow passed retention in history scavenger",
+				tag.Error(err),
+				tag.WorkflowNamespace(ns.Name().String()),
+				tag.WorkflowID(executionInfo.GetWorkflowId()),
+				tag.WorkflowRunID(mutableState.GetExecutionState().GetRunId()),
+			)
+		}
+	}
+	return nil
```

Comment:
Probably add a comment here says: Admin DeleteWorkflowExecution will attempt to delete history, but not guaranteed... (unless `Warning` returned in response is empty) so still need to explicitly delete history branch.

Actually the design of admin workflow deletion as a admin tool is: rely on history scavenger to clean up un-deleted history branches...

Reply:
I guess we will switch back to use history DeleteWorkflowExecution in next release?

Reply:
Yes. We can switch back to history DeleteWorkflowExecution in next release. I don't think a problem with admin delete. Even if the history fails to be deleted. The history scavenger will delete them in the next scan.